### PR TITLE
fix(madsim): drop the future within the scope of `TaskEnterGuard` when a task is aborted

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>

### CODE REVIEW

File: madsim/src/sim/task/mod.rs
Comment (kwannoel) [inline madsim/src/sim/task/mod.rs:273]: Ohhhh I see. The `continue` triggered us to directly proceed to the end of the loop, and call the destructor of runnable implicitly. Then it calls the drop for the struct, which spawned the thread.

Seems acceptable to me to drop it.

**Review by kwannoel — COMMENTED** [review]
LGTM once pass CI. @wangrunji0408 could you take a look too?

**Comment by kwannoel** [comment]
I considered opposite cases like:
```
    #[test]
    fn spawn_futures() {
        let runtime = madsim::runtime::Runtime::new();
        runtime.block_on(async move {
            let rt = Runtime::new().unwrap();
            let join_handle = rt.spawn(async move {
                Handle.spawn(async move {
                    madsim::time::sleep(Duration::from_secs(1)).await;
                    println!("hi");
                });
                Handle.spawn(async move {
                    madsim::time::sleep(Duration::from_secs(1)).await;
                    println!("hi2")
                });
            });
            madsim::time::sleep(Duration::from_secs(10)).await;
            join_handle.await.unwrap();
        });
        panic!("break");
    }
```

In such cases, we will always have task context, since their parent future will be executing with task context, and the spawning steps happen sequentially. The newly spawned futures will be handled by `Spawner`, which will schedule these tasks using its `sender`, and let the main event loop inside the executor execute these.

So I think we just need to handle the case of node kill / task abort as was done in this PR.

**Comment by kwannoel** [comment]
Many thanks @BugenZhao!

File: madsim/src/sim/task/mod.rs
Comment (BugenZhao) [inline madsim/src/sim/task/mod.rs:633]: Also relax this check so that `spawn` during drop by killing a node won't panic.

**Review by kwannoel — APPROVED** [review]

**Comment by kwannoel** [comment]
Let's also bump the version of tokio-postgres to v0.2.31 after this PR has been merged.

**Comment by BugenZhao** [comment]
Given that this PR has been tested in https://github.com/risingwavelabs/risingwave/pull/18852 and  RisingWave might be the largest project to adopt madsim, I believe we can safely merge this PR now.

**Pull request merged**
